### PR TITLE
Refactor AWS provisioners (resolves #1184, resolves #1185, resolves #1186, resolves #1306, resolves #1312, resolves #1325)

### LINKED ISSUE #1184: Move shared node managing functionality into base class
Right now the _removeNodes method in the AWS/CGCloud provisioners are almost identical to each other. Refactor them to move shared functionality into a base class. 

The entire setNodes function may be able to be shared between the classes.

### LINKED ISSUE #1185: Move rewritten CGCloud-lib code back into CGCloud
Especially `create_spot_instances`. Other methods like `retryBoto` and the IAM role policies may also make sense to move into cgcloud lib.

### LINKED ISSUE #1186: Eliminate unwieldy static methods from provisioner API
`AwsProvisioner(…).sshMaster("foo")` instead of `AwsProvisioner.ssh(AwsProvisioner.getLeaderIP())`. 
Possible solutions are
- partially initialized provisioner (make the provisioner constructor less demanding wrt to what it requires and only partially initialize the object depending on the paramters provided.
- change the provisioner API to be a factory for clusters: AwsProvisioner.getOrCreateCluster().sshLeader()

### LINKED ISSUE #1306: Retry on EC2 instance eventual consistency
Try to use cgcloud-lib code

### LINKED ISSUE #1312: Replace `retry_ec2` with `retry` in autoscaling code
`retry` does proper exponential back-off while `retry_ec2` is just a wrapper for backwards compatibility. Consider it deprecated.

Think twice about replacing the default delay and timeout values.

### LINKED ISSUE #1325: cluster_utils_test missing a dash in IAM profile name

### PR DESCRIPTION
**Pull request by cket**

### CODE REVIEW

**Comment by cket** [comment]
jenkins, test this please

**Comment by cket** [comment]
jenkins, test this please

**Comment by cket** [comment]
jenkins, test this please

**Comment by cket** [comment]
Jenkins, test this please

**Comment by cket** [comment]
jenkins, test this please

**Comment by cket** [comment]
jenkins, test this please

**Comment by cket** [comment]
jenkins, test this please

**Comment by cket** [comment]
@hannes-ucsc ready for review

File: src/toil/provisioners/__init__.py [hunk 1]
```diff
@@ -55,3 +34,41 @@ def _partialBillingInterval(instance):
...
+    impairedNodes = [node.id for node in nodes if statusMap.get(node.id, None) == 'impaired']
+    logger.warn('TOIL_AWS_NODE_DEBUG is set and nodes %s have failed EC2 status checks so '
+                'will not be terminated.', ' '.join(impairedNodes))
+    return healthyNodes
+
+
+class Cluster(object):
+    def __init__(self, clusterName, provisioner):
+        self.clusterName = clusterName
+        self.provisioner = None
```
Comment (hannes-ucsc) [inline src/toil/provisioners/__init__.py]: No point in this.

File: src/toil/provisioners/__init__.py [hunk 2]
```diff
@@ -55,3 +34,41 @@ def _partialBillingInterval(instance):
...
+        if provisioner == 'aws':
+            from toil.provisioners.aws.awsProvisioner import AWSProvisioner
+            self.provisioner = AWSProvisioner
+        elif provisioner == 'cgcloud':
+            from toil.provisioners.cgcloud.provisioner import CGCloudProvisioner
+            self.provisioner = CGCloudProvisioner
+        else:
+            raise RuntimeError('The only options are aws and cgcloud')
```
Comment (hannes-ucsc) [inline src/toil/provisioners/__init__.py]: I would `assert False, "Invalid provisioner '%s'" % provisioner`

File: src/toil/provisioners/abstractProvisioner.py [hunk 3]
```diff
@@ -60,7 +69,74 @@ def setNodeCount(self, numNodes, preemptable=False, force=False):
...
+            if log.isEnabledFor(logging.DEBUG):
+                for instance, nodeInfo in nodes:
+                    log.debug("Instance %s is about to be terminated. Its node info is %r. It "
+                              "would be billed again in %s minutes.", instance.id, nodeInfo,
+                              60 * self._remainingBillingInterval(instance))
+            instanceIds = [instance.id for instance, nodeInfo in nodes]
+        else:
+            # Without load info all we can do is sort instances by time left in billing cycle.
+            instances = sorted(instances,
+                               key=lambda instance: (self._remainingBillingInterval(instance)))
```
Comment (hannes-ucsc) [inline src/toil/provisioners/abstractProvisioner.py]: `key=self._remainingBillingInterval`

**Pull request merged**